### PR TITLE
Remove winmm support from NVDA

### LINKED ISSUE #2067: Audio output device names truncated to 32 characters
**Reported by HKatic on 2012-01-29 13:10**
In NVDA's synthesizer preferences dialog, the Output Device Combo Box displays incomplete device names. For example, I have 2 sound cards, the one which is Realteck High Definition Audio, and another which is SoundBlaster X-Fi Surround 5.1 Pro. However, they are displayed as follows:
Output device: Speakers (Realtek High Definiti
Output device: Speaker (Sound Blaster X-Fi Sur
This simply means that the entire device name doesn't fit into the Combo Box control, and that the control or dialog itself needs to be resized.

### LINKED ISSUE #16080: Remove the "use WASAPI for audio output" setting from the GUI
### Is your feature request related to a problem? Please describe.
The advanced settings GUI of NVDA is packed with some settings that are already widely tested by the community and per policy should not appear under advanced settings or in the GUI at all since their default behavior should not be changed by the user. Letting these hard settings changeable overcomplicates investigations especially when most significant issues were already fixed and the community already tested most use cases.
This applies also for use WASAPI setting, in most cases we have to find out who has manually changed this setting and who not. My understanding if the original implementation is that The users are expected to use WASAPI by default without being able to change it.

### Describe the solution you'd like
Remove the "use WASAPI for audio output" setting from the GUI and enforce WASAPI to enabled for everyone.

### Describe alternatives you've considered
Leave as is but we will never be able to discover all the issues that come with WASAPI since many people might use both and switch back and forth to compensate for issues deriving from WASAPI or nvWave. For true testing we should remove this setting completely from the GUI.

cc: @jcsteh what do you think? Are any drawbacks which come from only using WASAPI?

### PR DESCRIPTION
**Pull request by SaschaCowley**
### Link to issue number:

Closes #16080
Closes #2067 

### Summary of the issue:

The support for WASAPI (and Windows' Core Audio APIs more broadly) in NVDA is now quite mature, and maintenance of our winmm support is becoming untenable.

### Description of user facing changes:

The option to use WASAPI for audio output has been removed from NVDA's advanced settings. This has been on by default for some time, so it is unlikely to affect most users.

As the Mmdevice API does not have the concept of an ID to refer to the default output device, "Microsoft Sound Mapper" is no longer an option in the output device selection in Audio Settings. Users can instead choose "Default output device".

### Description of development approach

Removed all references to winmm in `nvwave.py`. Rewrote the device enumeration to use the Mmdevice API, via pycaw. Slightly modified the device selection logic in the settings GUI in light of Mmdevice not having an equivalent to Microsoft Sound Mapper.

### Testing strategy:

System and unit tests, as well as running NVDA, and changing output devices, including plugging and unplugging headphones while NVDA was running.

### Known issues with pull request:

None

### Code Review Checklist:

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [ ] API is compatible with existing add-ons.
- [x] Security precautions taken.

@coderabbitai summary

### CODE REVIEW

**Comment by SaschaCowley** [comment]
@coderabbitai please open an issue related to this one, for switching to using endpoint IDs rather than friendly names for identifying the selected output device in user configuration.

**Review by CyrilleB79 — CHANGES_REQUESTED** [review]
This PR seems incomplete:
* Various audio features (e.g. sound split, mute other apps, etc.) only work if WASAPI is enabled:
  * in settings dialog, the check of WASAPI to enable or disable controls for such settings should be removed (see (`nvwave.usingWasapiWavePlayer` calls)
  * in the User Guide, the documentation of these settings mention WASAPI requirement; this should be removed
* In config spec, `WASAPI` should be removed from `audio` section, and this removal should be documented in API-breaking changes. Or in case you want to keep it as a no-effect parameter, you should at least document it in the change log and remove all of its usage in NVDA's code.
* I also wonder if the removal of `AdvancedPanelControls.wasapiComboBox` should be documented in the API-breaking changes.

**Comment by SaschaCowley** [comment]
@CyrilleB79 thanks for the thorough notes. I believe I've addressed all of your points now.

File: source/nvwave.py [hunk 1]
```diff
@@ -1107,12 +569,12 @@ def _isDefaultDevice(cls, name):
 		return name == next(_getOutputDevices())[1]
 
 
+WavePlayer = WasapiWavePlayer
+fileWavePlayer: Optional[WavePlayer] = None
+fileWavePlayerThread = None
```
Comment (seanbudd) [inline source/nvwave.py:574]: can this get a type hint?

File: source/nvwave.py [hunk 2]
```diff
@@ -173,477 +99,21 @@ class AudioPurpose(Enum):
...
+	for i in range(endpointCollection.GetCount()):
+		device = AudioUtilities.CreateDevice(endpointCollection.Item(i))
+		# This should never be None, but just to be sure
+		if device is not None:
+			yield device.id, device.FriendlyName
+		else:
+			continue
 
 
 def getOutputDeviceNames():
```
Comment (seanbudd) [inline source/nvwave.py:119]: can this get a return type added?

File: source/nvwave.py [hunk 3]
```diff
@@ -173,477 +99,21 @@ class AudioPurpose(Enum):
...
-			# device will be closed and _waveout set to None,
-			# triggering re-open.
-			self._handleWinmmError(message=messageOnFailure)
-			return False
-
-
-WavePlayer = WinmmWavePlayer
-
-
 def _getOutputDevices():
```
Comment (seanbudd) [inline source/nvwave.py:102]: can this get a return type?

**Review by seanbudd — COMMENTED** [review]
Thanks @SaschaCowley - mostly looks good to me!

File: source/nvwave.py [hunk 4]
```diff
@@ -99,9 +100,9 @@ class AudioPurpose(Enum):
 	SOUNDS = auto()
 
 
-def _getOutputDevices():
-	"""Generator, returning device ID and device Name in device ID order.
-	@note: Depending on number of devices being fetched, this may take some time (~3ms)
+def _getOutputDevices() -> Iterator[tuple[str, str]]:
```
Comment (seanbudd) [inline source/nvwave.py:103]: I'd suggest using the Generator type as their handling can be different

```suggestion
def _getOutputDevices() -> Generator[tuple[str, str], None, None]:
```

File: source/nvwave.py [hunk 5]
```diff
@@ -116,35 +117,30 @@ def _getOutputDevices():
...
-	@rtype: str
+	:param ID: The device ID.
+	:return: The device name.
 	"""
 	device = AudioUtilities.GetDeviceEnumerator().GetDevice(id)
 	return AudioUtilities.CreateDevice(device).FriendlyName
 
 
-def outputDeviceNameToID(name: str, useDefaultIfInvalid=False) -> int:
+def outputDeviceNameToID(name: str, useDefaultIfInvalid: bool = False) -> str:
```
Comment (seanbudd) [inline source/nvwave.py:137]: as this return type is changing, can this be listed as an API breaking change?

**Review by seanbudd — APPROVED** [review]

**Review by seanbudd — APPROVED** [review]
Thanks @SaschaCowley

**Review by CyrilleB79 — APPROVED** [review]

**Review by Qchristensen — APPROVED** [review]
Reads well, good work.

**Comment by Adriani90** [comment]
Thanks Sascha this is very appreciated cleanup.

**Comment by CyrilleB79** [comment]
@SaschaCowley, after this PR, if I plug an USB headset, but select my PC speaker for NVDA audio output, the sound still comes in the headset rather than from PC sdpeakers.

**Comment by Adriani90** [comment]
@SaschaCowley it seems the NVWave advanced logging channel is still there, I mean the one introduced in #11582  and  #11614. Does it have to be removed as well? Or is it not related?

**Pull request merged**